[FEATURE-BRANCH] ImageField: add support to new fields of type image

jfcalvo opened the pull request:
# Description

This PR is the first one related with the addition of a new field of type `image` including the following changes:
* Changes in the backend to support a new `image` field type supporting URLs and Data URLs.
* Changes in the frontend rendering this new `image` field for records.
* Changes in the SDK to define datasets with `image` fields and create records specifying images.
* Changes in the documentation with examples on how to use the SDK with this new type of field.

This new type will have the following characteristics:
* A new filed type `image` is supported (without additional settings).
* Records created with fields of type `image` are validated to have correct Web URLs or data URLs like:
  *  `https://argilla.io/image.jpeg`:
     * `HTTPS` schema is present
     * A host and a path is present.
     * The URL is not longer than `2038` characters.
  * `http://argilla.io/image.jpeg`:
     * `HTTP` schema is present
     * A host and a path is present.
     * The URL is not longer than `2038` characters.
  * `data:image/webp;base64,UklGRhgCAABXRUJQVlA4WA`:
     * `data` schema is present
     * A path is present.
     * The Data URL is not longer than `5` million characters.
     * A valid MIME type is present.

Closes #5276

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Adding new tests to our suite checking this new field.

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

jfcalvo committed:
feat: add support to new fields of type image

jfcalvo committed:
chore: update CHANGELOG.md

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

jfcalvo committed:
improvement: validate that records values with associated image fields has correct URLs

jfcalvo committed:
chore: update CHANGELOG.md

jfcalvo committed:
chore: improve import

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

frascuchon committed:
[FEATURE] `argilla server`: Add support to image fields on search engine (#5355)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR adds support to indexing datasets and records with image fields.

Note: A datasets reindex is required with this change.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

jfcalvo committed:
feat: add length validation for record image field values

jfcalvo committed:
Merge branch 'feat/image-field' into feat/add-image-field-support

jfcalvo committed:
feat: add MIME type validation for records using image fields values with Data URLs

frascuchon commented on argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py:
```diff
@@ -355,6 +355,35 @@ async def test_create_dataset_records_bulk_with_image_field_web_url_exceeding_ma
 
         assert (await db.execute(select(func.count(Record.id)))).scalar_one() == 0
 
+    async def test_create_dataset_records_bulk_with_image_field_data_url_using_invalid_mime_type(
```
Do we have another test to check the accepted mime types?

frascuchon committed:
[FEATURE] `argilla`: add support to create image fields (#5372)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR adds support to work with image fields using the SDK. 

## TODO (as separate PRs)

- [ ] Add documentation section
- [ ] Verify behavior using from/to_hub methods
- [ ] Add tooling to read data URLs from image files/folders (based on
logic defined in `image_to_html`)
- [ ] Add some media type validation (supported image types) 
- [ ] [Optional] Add tooling to re-scale images. The lower the image
sizes, the better the UI work. We can define a transparent re-scaling
process to support large images (currently the limit size in the backend
is 5MB for data URL)


**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: José Francisco Calvo <jose@argilla.io>

davidberenstein1957 commented on argilla-server/src/argilla_server/search_engine/commons.py:
```diff
@@ -150,6 +150,13 @@ def es_mapping_for_field(field: Field) -> dict:
+            }
+        }
     else:
         raise Exception(f"Index configuration for field of type {field_type} cannot be generated")
```
```suggestion
    if field.is_text:
        return {es_field_for_record_field(field.name): {"type": "text"}}
    elif field.is_image:
        return {
            es_field_for_record_field(field.name): {
                "type": "object",
                "enabled": False,
            }
        }
    else:
        raise Exception(f"Index configuration for field of type {field_type} cannot be generated")
```

davidberenstein1957 commented on argilla-server/src/argilla_server/validators/records.py:
```diff
@@ -25,6 +28,19 @@
 from argilla_server.models import Dataset, Record
 
+IMAGE_FIELD_WEB_URL_MAX_LENGTH = 2038
+IMAGE_FIELD_DATA_URL_MAX_LENGTH = 5_000_000
```
I think we also added this validation to the python client, can your remove it from there?

burtenshaw replied:
There seem to be a number of validators duplicated between server and sdk. I think it makes sense for them to stand alone.

davidberenstein1957 commented on argilla-server/src/argilla_server/validators/records.py:
```diff
@@ -71,6 +88,54 @@ def _validate_extra_fields(self, dataset: Dataset, fields: Dict[str, str]) -> No
+        self, field_name: str, field_value: str, parse_result: Union[ParseResult, ParseResultBytes]
+    ) -> None:
+        if not parse_result.netloc or not parse_result.path:
+            raise UnprocessableEntityError(f"image field {field_name!r} has an invalid URL value")
```
should we rename "image field" to ImageField in error messaging?

burtenshaw replied:
We use lower case of class name in all the messages.

davidberenstein1957 commented on argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py:
```diff
@@ -214,3 +217,198 @@ async def test_create_dataset_records(
+
+        assert response.status_code == 422
+        assert response.json() == {
+            "detail": f"record at position 0 is not valid because image field 'image' value is exceeding the maximum length of 5000000 characters for Data URLs",
```
@burtenshaw perhaps the record position is a very inconvenient message, could return the actual record along with it? Additionally, I think it would be nice to update the `__repr__` to avoid showing the entire vector on printing.

davidberenstein1957 replied:
If you agree, I will create some issues.

burtenshaw replied:
I think it makes a big difference if record 0 or any other record raises the error in debugging. That said, these error messages could be improved with a representation of the record.

davidberenstein1957 replied:
@burtenshaw, I agree with showing the index but I think being able to see the record would be nice too.

davidberenstein1957 commented on argilla-server/tests/unit/api/handlers/v1/fields/test_update_field.py:
```diff
@@ -0,0 +1,54 @@
+
+
+@pytest.mark.asyncio
+class TestUpdateField:
```
should we test updates with wrong values?

burtenshaw replied:
We're testing validations which should be the same thing.

burtenshaw committed:
Update argilla-server/src/argilla_server/search_engine/commons.py

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

burtenshaw committed:
Merge branch 'develop' into feat/add-image-field-support

burtenshaw committed:
chore: update changelog

burtenshaw committed:
fix: merged error self >> cls

burtenshaw committed:
fix: merged error self >> cls

burtenshaw committed:
Merge branch 'feat/add-image-field-support' of https://github.com/argilla-io/argilla into feat/add-image-field-support

burtenshaw committed:
fix: merge mistake - suggestions at self not record

burtenshaw committed:
fix: merge mistake: more uncaught refactoring to class methods

damianpumar committed:
🌅 Image field (#5389)

This PR close this issue:
https://github.com/argilla-io/argilla/issues/5278

---------

Co-authored-by: Leire Aguirre <leire@argilla.io>

leiyre committed:
update image placeholder

burtenshaw committed:
docs: add api reference for image field

burtenshaw committed:
docs: add how to guide for image field

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
[BUGFIX] map all field types in record mapper (#5408)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR changes the mapper to identify fields of all types (not just
text).

Closes #5406 

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
[BUGFIX] validate iterable dataset in log method (#5410)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

Closes #<issue_number>

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

jfcalvo committed:
Merge branch 'develop' into feat/add-image-field-support

jfcalvo committed:
Merge branch 'develop' into feat/add-image-field-support

frascuchon commented on argilla/CHANGELOG.md:
```diff
@@ -36,6 +36,10 @@ These are the section headers that we use:
+### Added
+
+- Added test mock for ArgillaAPI validation step so that the validation method is mocked instead of the entire API.
+
```
The CHANGELOG should include only changes affecting the final users. Adding tests or internal refactors shouldn't affect the project changelog.

```suggestion
```

burtenshaw committed:
fix: rely on AbstractField __init__ for text field as implemented in branch

jfcalvo committed:
Update argilla/CHANGELOG.md

Co-authored-by: Paco Aranda <frascuchon@gmail.com>

frascuchon commented on argilla/docs/reference/argilla/records/records.md:
````diff
@@ -23,6 +23,21 @@ dataset.records.log(
+)
+```
+
+1. The image can be referenced as either a remote url or a data uri.
````
Adding an example using data URI would be great @sdiazlor

frascuchon replied:
@burtenshaw  maybe we can include here the `image_to_data_uri` (or whatever) util function usage once you have implemented it

damianpumar committed:
🐛 Show use markdown just for text type

frascuchon commented on argilla/docs/how_to_guides/image_field.md:
```diff
@@ -0,0 +1,56 @@
+            fields={"image": "https://example.com/image.jpg"},
+        ),
+        rg.Record(
+            fields={"image": "data:image/png;base64,iV..."}, # (1)
```
same here about the imate-to-data-uri function

frascuchon committed:
Merge branch 'develop' into feat/add-image-field-support

damianpumar committed:
✨ Add badge

damianpumar committed:
Merge branch 'feat/add-image-field-support' of https://github.com/argilla-io/argilla into feat/add-image-field-support

damianpumar committed:
✨ Add overflow

burtenshaw committed:
[FEATURE] cast datasets pil objects to base64 in SDK (#5433)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->
This PR introduces a utility into the SDK so that it automatically casts
PIL objects to base64 based on the datasets `Features`

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

jfcalvo committed:
Merge branch 'develop' into feat/add-image-field-support

davidberenstein1957 committed:
docs: 5405 docs tutorial on the usage of image fields (#5413)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

Closes #5405 

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: burtenshaw <ben@argilla.io>
Co-authored-by: Paco Aranda <frascuchon@gmail.com>

jfcalvo committed:
Merge branch 'develop' into feat/add-image-field-support

frascuchon committed:
chore: Update changelog

frascuchon committed:
Merge branch 'feat/add-image-field-support' of github.com:argilla-io/argilla into feat/add-image-field-support

frascuchon committed:
ci: Restore to develop server tag
